lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
   void getAnalysisUsage(AnalysisUsage &AU) const override {
     AU.addRequired<BlockFrequencyInfoWrapperPass>();
   }
-
 };
 }
 
 char StructFieldCacheAnalysisPass::ID = 0;
 INITIALIZE_PASS_BEGIN(StructFieldCacheAnalysisPass, "struct-field-cache-analysis", "Struct Field Cache Analysis", false, false)
+INITIALIZE_PASS_DEPENDENCY(BlockFrequencyInfoWrapperPass)
 INITIALIZE_PASS_END(StructFieldCacheAnalysisPass, "struct-field-cache-analysis", "Struct Field Cache Analysis", false, false)
 ModulePass *llvm::createStructFieldCacheAnalysisPass() { return new StructFieldCacheAnalysisPass; }
 
-static bool performStructFieldCacheAnalysis(Module &M)
+namespace llvm{
+typedef unsigned FieldNumType;
+
+class StructFieldAccessInfo;
+class StructFieldAccessManager
+{
+  /* This class is used to keep track of all StructFieldAccessInfo objects in the program
+     and make sure only one StructFieldAccessInfo object for each type of struct declared
+     in the program.
+  */
+ public:
+  enum StructDefinitionType{
+    GlobalStruct,
+    GlobalStructPtr,
+    GlobalStructPtrPtr,
+    LocalStruct,
+    LocalStructPtr,
+    LocalStructPtrPtr
+  };
+
+  StructFieldAccessManager(const Module& M, function_ref<BlockFrequencyInfo *(Function &)> L):
```

Comment:
Fixed. I also used a better name is later branches when more functions are needed...